code/game/objects/items/robot/robot_upgrades.dm

```diff
@@ -381,3 +381,22 @@
 	R.module.add_module(S, FALSE, TRUE)
 
 	return 1
+
+/obj/item/borg/upgrade/ai
+	name = "B.O.R.I.S. module"
+	desc = "Bluespace Optimized Remote Intelligence Synchronization. An uplink device which takes the place of an MMI in cyborg endoskeletons, creating a robotic shell controlled by an AI."
+	icon_state = "boris"
+	origin_tech = "engineering=4;magnets=4;programming=4"
+
+/obj/item/borg/upgrade/ai/action(mob/living/silicon/robot/R)
+	if(..())
+		return
+	if(R.shell)
+		to_chat(usr, "<span class='warning'>This unit is already an AI shell!</span>")
+		return
+	if(R.key) //You cannot replace a player unless the key is completely removed.
+		to_chat(usr, "<span class='warning'>Intelligence patterns detected in this [R.braintype]. Aborting.</span>")

```

Comment:
It's just a minor nitpick of mine, but you could do an early return here so you don't have do make an else down here.